Add ability to write/read using coordinates (incl. tests)

ucyo opened the pull request:
I added the ability to write and read from a netcdf file using the coordinate positions. The C functions `nc_get_var1_<type>` and `and nc_put_var1_<type>` are being used. This concludes all supported methods in the C drivers writing directly to a dataset.

ucyo committed:
Add support for index-based read and write

ucyo committed:
Add commentary about types

ucyo committed:
Add UnravelIndex and product functions for testing purposes of index based read/write

ucyo committed:
Add tests for read values based on index

ucyo committed:
Add changes to generate.go and execute ones to generate appropiate functions

ucyo committed:
Write wrapper for testing idx selection and write

ucyo committed:
Use only positive values for shape
  - With this LenDims() can be used directly

ucyo committed:
Typo

ucyo committed:
Initial .gitignore file

ucyo committed:
Don;t forget the milk

ucyo committed:
Version bump and Travis run

fhs commented on TODO.md:
```diff
@@ -0,0 +1,2 @@
+# TODO
+- Use Pointer in method declaration instead of actual Variable [see](https://dave.cheney.net/2016/03/19/should-methods-be-declared-on-t-or-t)
```
Since the structs are small with immutable fields, value receiver makes in this case. But you're right that it needs to be justified with profiling. See https://github.com/golang/go/wiki/CodeReviewComments#receiver-type

ucyo replied:
I'd also leave it as is until we have some benchmarks with bigger netCDF files, but we should keep it in mind...

fhs commented on netcdf/netcdf.go:
```diff
@@ -37,3 +37,36 @@ func okData(a typedArray, t Type, n int) error {
+	var ndim = len(shape)
+	var coord = make([]uint64, ndim)
+
+	for _, v := range shape {
```
move this loop to the beginning of the function

fhs commented on netcdf/nc_double.go:
```diff
@@ -55,6 +55,20 @@ func (a Attr) ReadFloat64s(val []float64) (err error) {
 }
 
+// ReadIdxFloat64 returns a value via index position
+func (v Var) ReadIdxFloat64(idx []uint64) (val float64, err error) {
```
I'd rename this method to `ReadFloat64At`, which is similar to os.File ReadAt and https://godoc.org/gonum.org/v1/gonum/mat#Dense.At

fhs commented on netcdf/nc_double.go:
```diff
@@ -55,6 +55,20 @@ func (a Attr) ReadFloat64s(val []float64) (err error) {
+}
+
+// WriteIdxFloat64 sets a value via its index position
+func (v Var) WriteIdxFloat64(idx []uint64, val float64) (err error) {
```
similarly, `WriteFloat64At`

fhs commented on netcdf/datasets_test.go:
```diff
@@ -384,3 +385,217 @@ func TestVersion(t *testing.T) {
+	}
+
+	for _, test := range results {
+		var result = product(test.shape)
```
nitpick: use := and maybe a shorter name

    if p := product(test.shape); p != test.expected {

fhs commented on netcdf/datasets_test.go:
```diff
@@ -384,3 +385,217 @@ func TestVersion(t *testing.T) {
+	}
+
+	for _, test := range tests {
+		var result, err = UnravelIndex(test.idx, test.shape)
```
use :=

fhs commented on netcdf/nc_double.go:
```diff
@@ -92,7 +106,46 @@ func testReadFloat64s(v Var, n uint64) error {
+	for i := 0; i < int(n); i++ {
+		shape, _ := v.LenDims()
+		coords, _ := UnravelIndex(uint64(i), shape)
+		expected := float64(i + 10)
```
`expected := data[i]` and replace `data[i]` below with `expected`

fhs commented on netcdf/nc_double.go:
```diff
@@ -92,7 +106,46 @@ func testReadFloat64s(v Var, n uint64) error {
+		}
+		v.WriteIdxFloat64(coord, float64(i))
+	}
+	for i := 0; i < ndim; i++ {
```
This loop should be in `testReadFloat64Idx` but then `testReadFloat64Idx` already does something similar

ucyo replied:
Kind of... I have to test what has been written. So I put the `ReadFloat64Idx` into the `Write` instead of the other way around

fhs replied:
maybe add a comment explaining why testing Read there

fhs commented on netcdf/datasets_test.go:
```diff
@@ -384,3 +385,217 @@ func TestVersion(t *testing.T) {
+		t.Fatalf("Close failed: %v\n", err)
+	}
+}
+func TestIdx(t *testing.T) {
```
add space between functions

ucyo replied:
Dunno how `fmt` missed this...

fhs reviewed (changes requested):
Thanks for working on this!

ucyo committed:
Made changes according to change requests of PR #7

  - Renamed functions
  - Cleanup of code
  - Reordering of for loops and input checks

ucyo commented:
Thanks for the comments. I applied the changes except the one for the `testWriteFloat64Idx` since the written data needs to be checked and which function gets nested into which shouldn't matter

fhs commented on netcdf/datasets_test.go:
```diff
@@ -384,3 +385,216 @@ func TestVersion(t *testing.T) {
+		t.Fatalf("Close failed: %v\n", err)
+	}
+}
+func TestAt(t *testing.T) {
```
still missing a space ☺

ucyo replied:
Apparently `fmt` does not check for that .. :confused:

ucyo committed:
Cleanup

fhs commented:
Merged. Thanks for all the changes.
